### PR TITLE
Add provision watcher implementation

### LINKED ISSUE #30: Add settings support for provisionwatchers
Dynamically create provisionwatcher instances based on configuration settings so that device services can implement dynamic discovery.

For more details see the [Device Services functional requirements](https://wiki.edgexfoundry.org/download/attachments/7602423/edgex-cali-delhi-device-requirements-v8.pdf?version=1&modificationDate=1533761857000&api=v2) document.

**Estimated effort**: 3d

### PR DESCRIPTION
**Pull request by hahattan**
Add provision watcher implementation refer to #204, change some naming to be consistent with latest SDK and add unit tests on `ProvisionWatcherCache`

Fix #30

### CODE REVIEW

**Comment by iain-anderson** [comment]
I don't think we should include population of the PWs from the configuration file. We do it for Devices and many people find that the behavior isn't what they expect. (Also, PWs are metadata, not configuration)

**Comment by hahattan** [comment]
> I don't think we should include population of the PWs from the configuration file. We do it for Devices and many people find that the behavior isn't what they expect. (Also, PWs are metadata, not configuration)

I made a new commit to load the PW from metadata instead of configuration (based on your comment and the new discovery legacy design)

**Comment by cloudxxx8** [comment]
@iain-anderson Is the current approach fine with you?

File: internal/cache/provisionwatchers.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	defer p.mutex.Unlock()
+
+	if watcher, ok := p.pwMap[name]; ok {
+		return *watcher, ok
+	} else {
+		return contract.ProvisionWatcher{}, ok
+	}
+}
+
+// ForId returns a provisionwatcher with the given device id.
```
Comment (iain-anderson) [inline internal/cache/provisionwatchers.go]: Not "device id". Other than that, LGTM

Comment (hahattan) [reply internal/cache/provisionwatchers.go]: Thanks I'll fix this.

**Comment by hahattan** [comment]
fixed the request change and rebased.

**Pull request merged**